cpp/src/arrow/flight/client.cc

```diff
@@ -569,6 +569,36 @@ Status FlightClient::DoAction(const FlightCallOptions& options, const Action& ac
   return DoAction(options, action).Value(results);
 }
 
+arrow::Result<std::unique_ptr<ActionCancelFlightInfoResult>>
+FlightClient::CancelFlightInfo(const FlightCallOptions& options, const FlightInfo& info) {
+  ARROW_ASSIGN_OR_RAISE(auto body, info.SerializeToString());
+  Action action{ActionType::kCancelFlightInfo.type, Buffer::FromString(body)};
+  ARROW_ASSIGN_OR_RAISE(auto stream, DoAction(options, action));
+  ARROW_ASSIGN_OR_RAISE(auto result, stream->Next());
```

Comment:
We should drain the rest of the DoAction stream since that's the only way we can get the server error (if any). You can see how this is done in the Flight SQL codebase: https://github.com/apache/arrow/blob/766e25440250dde9117e19245389badb5ed7678c/cpp/src/arrow/flight/sql/client.cc#L131-L137

Reply:
I think `ReadResult` can only be in an `_internal.h` header because it uses Protobuf. But `DrainResultStream` could become `ResultStream::Drain`.

Reply:
@lidavidm ping

Reply:
Ah, sorry.

The `Any` was a choice by the original Dremio developers, it gives you more type safety in Protobuf but is largely redundant, yes. (That said, even when wrapped in Any, we can still provide a function/method to parse it right?)

Reply:
But I would prefer not to wrap in Any, so we can keep it like this.

Reply:
> even when wrapped in Any, we can still provide a function/method to parse it right?

Right.

> But I would prefer not to wrap in Any, so we can keep it like this.

OK! Thanks!